virtool/ws/server.py

```diff
@@ -78,6 +78,24 @@ def remove_connection(self, connection: WSConnection):
         except ValueError:
             pass
 
+    async def close_connections_with_session_id(self, session_id):
+        """
+        Closes connections associated with a session_id
+        """
+        for connection in self._connections:
+            if connection.session_id == session_id:
+                await connection.close(1000)
+
+    async def close_expired_websocket_connections(self):
+        """
+        Closes connections with expired sessions.
+        """
+        while True:
+            for connection in self._connections:
+                if not connection.user_id:
+                    await connection.close(1001)
```

Comment:
You'll need to check if the session is valid to determine if it should be expired. Query the `sessions` data_layer with the `session_id` to check if the session is valid. Keep in mind the session must be authenticated to be valid for a WS connection.